common/src/chain/transaction/signature/sighash/hashable.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    chain::{
+        signature::{sighashtype, TransactionSigError},
+        TxInput, TxOutput,
+    },
+    primitives::id::{hash_encoded_to, DefaultHashAlgoStream},
+};
+
+pub trait SignatureHashableElement {
+    fn signature_hash(
+        &self,
+        stream: &mut DefaultHashAlgoStream,
+        mode: sighashtype::SigHashType,
+        target_input: &TxInput,
+        target_input_num: usize,
+    ) -> Result<(), TransactionSigError>;
+}
+
+impl SignatureHashableElement for &[TxOutput] {
+    fn signature_hash(
+        &self,
+        stream: &mut DefaultHashAlgoStream,
+        mode: sighashtype::SigHashType,
+        _target_input: &TxInput,
+        target_input_num: usize,
+    ) -> Result<(), TransactionSigError> {
+        match mode.outputs_mode() {
+            sighashtype::OutputsMode::All => {
+                hash_encoded_to(self, stream);
+            }
+            sighashtype::OutputsMode::None => (),
+            sighashtype::OutputsMode::Single => {
+                let output = self.get(target_input_num).ok_or({
+                    TransactionSigError::InvalidInputIndex(target_input_num, self.len())
+                })?;
+                hash_encoded_to(&output, stream);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SignatureHashableInputs<'a> {
+    inputs: &'a [TxInput],
+    /// Include utxos of the inputs to make it possible to verify the inputs scripts and amounts without downloading the full transactions
+    inputs_utxos: &'a [&'a TxOutput],
+}
```

Comment:
Not sure if it's an improvement but using `&[(TxInput, &TxOutput)]` enforces the same number of inputs and outputs by construction. Enforcing that in the constructor as done now is fine too.

Reply:
Your solution is nicer, but the only issue with it is that usually inputs come in a Vec straight from a transaction, which will mean that creating this Vec/Slice of tuples will require rearranging the inputs every time we need a sighash, since it's not natural to have this format of an array.